stylo: Improve restyling performance

emilio opened the pull request:
This commit adds hooks to the Servo style traversal to avoid traversing all the
DOM for every restyle. Additionally it changes the behavior of the dirty flag to
be propagated top down, to prevent extra overhead when an element is dirtied.

This commit doesn't aim to change the behavior on Servo just yet, since Servo does extra job when dirtying the node related with DOM revision counters that might be necessary.

CC @asajeffrey for the DOM revision counters stuff. When a node is dirty, do all its descendants really need to increment the revision counter, or is this an unintended effect? My intuition is that this is hurting performance quite a lot for servo.

r? @bholley 

<!-- Please describe your changes on the following line: -->

---

<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `__` with appropriate data: -->
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors

<!-- Either: -->
- [x] These changes do not require tests because no geckolib tests yet.

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/12563)

<!-- Reviewable:end -->

emilio committed:
stylo: Don't traverse the whole dom every restyle, propagate the dirty flag down the DOM.

This commit adds hooks to the Servo style traversal to avoid traversing all the
DOM for every restyle. Additionally it changes the behavior of the dirty flag to
be propagated top down, to prevent extra overhead when an element is dirtied.

This commit doesn't aim to change the behavior on Servo just yet, since Servo
might rely on a full bottom up reconstruction of the flows. I'll need to double
check and implement that separately.

emilio committed:
stylo: don't double check that the root node is dirty in restyle_tree, since now we ensure it in the caller.

highfive commented:
Heads up! This PR modifies the following files:
 * @bholley: components/style/traversal.rs, components/style/sequential.rs, components/style/parallel.rs

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify **unsafe code**. Please review it carefully!
* These commits modify style code, but no tests are modified. Please consider adding a test!

emilio committed:
style: Avoid propagating the restyle flag all through the dom when a node gets dirty.

This puts us in pair with stylo.

emilio commented:
@bors-servo: try

bors-servo commented:
:hourglass: Trying commit 31131d6 with merge c8e2fc8...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/2183)

highfive commented:
▶ FAIL [expected PASS] /css-transforms-1_dev/html/perspective-containing-block-dynamic-1a.htm
      └   → /css-transforms-1_dev/html/perspective-containing-block-dynamic-1a.htm f6da2d5fe045bd1531726e432bd3266de6f8c5c3
    /css-transforms-1_dev/html/reference/containing-block-dynamic-1-ref.htm 01beec2c279d618e0b4187715ab4f5be65d90eec
    Testing f6da2d5fe045bd1531726e432bd3266de6f8c5c3 == 01beec2c279d618e0b4187715ab4f5be65d90eec
    
      ▶ FAIL [expected PASS] /css21_dev/html4/at-charset-utf16-be-002.htm
      └   → /css21_dev/html4/at-charset-utf16-be-002.htm 1f9b02c0db24580a86d61ff094ede9b88b000a38
    /css21_dev/html4/reference/at-charset-ref.htm ac193431243c865f246dc0c7973667ace7689f93
    Testing 1f9b02c0db24580a86d61ff094ede9b88b000a38 == ac193431243c865f246dc0c7973667ace7689f93
    
      ▶ FAIL [expected PASS] /css21_dev/html4/dynamic-top-change-002.htm
      └   → /css21_dev/html4/dynamic-top-change-002.htm 6a7e222474d15bce31082495376e468d9a87c443
    /css21_dev/html4/reference/dynamic-top-change-002-ref.htm 9839e5fd45dc0e2be1cc23b3c2c2d0a4c1d85f99
    Testing 6a7e222474d15bce31082495376e468d9a87c443 == 9839e5fd45dc0e2be1cc23b3c2c2d0a4c1d85f99
    
      ▶ FAIL [expected PASS] /css21_dev/html4/dynamic-top-change-003.htm
      └   → /css21_dev/html4/dynamic-top-change-003.htm 6a7e222474d15bce31082495376e468d9a87c443
    /css21_dev/html4/reference/dynamic-top-change-002-ref.htm 9839e5fd45dc0e2be1cc23b3c2c2d0a4c1d85f99
    Testing 6a7e222474d15bce31082495376e468d9a87c443 == 9839e5fd45dc0e2be1cc23b3c2c2d0a4c1d85f99
    
      ▶ FAIL [expected PASS] /css21_dev/html4/dynamic-top-change-004.htm
      └   → /css21_dev/html4/dynamic-top-change-004.htm 435beb3bd5258fc918ac2e27b4c22a0875789b4d
    /css21_dev/html4/reference/dynamic-top-change-002-ref.htm 9839e5fd45dc0e2be1cc23b3c2c2d0a4c1d85f99
    Testing 435beb3bd5258fc918ac2e27b4c22a0875789b4d == 9839e5fd45dc0e2be1cc23b3c2c2d0a4c1d85f99
    
      ▶ FAIL [expected PASS] /css21_dev/html4/first-letter-dynamic-003a.htm
      └   → /css21_dev/html4/first-letter-dynamic-003a.htm 05c3e7d5e7f9e20bc71a2992df75739754bcd7dd
    /css21_dev/html4/reference/first-letter-dynamic-003-ref.htm fe25de5a57e730bd2965c6f7a52d7255c2e01e7a
    Testing 05c3e7d5e7f9e20bc71a2992df75739754bcd7dd == fe25de5a57e730bd2965c6f7a52d7255c2e01e7a
    
      ▶ FAIL [expected PASS] /css21_dev/html4/first-letter-dynamic-003b.htm
      └   → /css21_dev/html4/first-letter-dynamic-003b.htm 55de50ac439243ffa380041ceabb912199b8720f
    /css21_dev/html4/reference/first-letter-dynamic-003-ref.htm fe25de5a57e730bd2965c6f7a52d7255c2e01e7a
    Testing 55de50ac439243ffa380041ceabb912199b8720f == fe25de5a57e730bd2965c6f7a52d7255c2e01e7a
    
      ▶ FAIL [expected PASS] /css21_dev/html4/run-in-text-between-003.htm
      └   → /css21_dev/html4/run-in-text-between-003.htm 7a880be74d2ad2d45520cea8cca05bd58b772752
    /css21_dev/html4/reference/run-in-text-ref.htm 638e3ddce4ea49b428602b944937bb27f420bd63
    Testing 7a880be74d2ad2d45520cea8cca05bd58b772752 == 638e3ddce4ea49b428602b944937bb27f420bd63
    
      ▶ FAIL [expected PASS] /css21_dev/html4/z-index-dynamic-001.htm
      └   → /css21_dev/html4/z-index-dynamic-001.htm 7d53c98773aae0945c22d7e5bbe05999f7ccd0c7
    /css21_dev/html4/reference/ref-filled-green-100px-square.htm 82d94d2027e216f7a2d31fcc53afb5e048516cef
    Testing 7d53c98773aae0945c22d7e5bbe05999f7ccd0c7 == 82d94d2027e216f7a2d31fcc53afb5e048516cef

emilio commented:
I *think* these are servo incremental layout bugs uncovered by this patch (since right now we relayout way less than before), I'll verify this during the weekend.

emilio commented:
What a better way to know?

@bors-servo: try

bors-servo commented:
:hourglass: Trying commit 99cb3ae with merge 6368273...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/2184)

emilio commented:
@bors-servo: try

bors-servo commented:
:hourglass: Trying commit 01a8d96 with merge 37d2095...

emilio commented:
@bors-servo: try- retry clean

emilio commented:
@bors-servo: try

bors-servo commented:
:hourglass: Trying commit 956f09e with merge 0fcd30c...

bors-servo commented:
:broken_heart: Test failed - [linux-dev](http://build.servo.org/builders/linux-dev/builds/2171)

emilio commented:
@bors-servo: try

bors-servo commented:
:hourglass: Trying commit aa90f7b with merge cda51f8...

bors-servo commented:
:broken_heart: Test failed - [linux-dev](http://build.servo.org/builders/linux-dev/builds/2172)

emilio commented:
Sorry for the spam, the linux builders failed because there was a typo in the geckolib traversal.rs file, but I wanted to look at the [mac results](http://build.servo.org/builders/mac-rel-css/builds/2187/steps/test/logs/stdio), which are more interesting.

emilio commented:
@bors-servo: try

I think I've got a few (all?) of them, was a super-stupid thing, layout only checked for `has_dirty_descendants`, which in practice **before** it meant `is_dirty || has_dirty_descendants` due to how the flags were propagated.

bors-servo commented:
:hourglass: Trying commit 7be7642 with merge 560629b...

emilio commented:
@bors-servo: try- retry clean

emilio commented:
@bors-servo: try

bors-servo commented:
:hourglass: Trying commit eab954a with merge ba40dbc...

bors-servo commented:
:hourglass: Trying commit eab954a with merge a6ca1e0...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/2196)

highfive commented:
▶ FAIL [expected PASS] /css21_dev/html4/at-charset-072.htm
      └   → /css21_dev/html4/at-charset-072.htm ef6b783123e594032837dab6a6c54e635a08c07e
    /css21_dev/html4/reference/ref-green-background.htm e94b172fccc7ab83de085eab014d69f44f286570
    Testing ef6b783123e594032837dab6a6c54e635a08c07e == e94b172fccc7ab83de085eab014d69f44f286570
    
      ▶ FAIL [expected PASS] /css21_dev/html4/media-dependency-003.htm
      └   → /css21_dev/html4/media-dependency-003.htm aa3afd43bed732a0c6cdc22b3a71303a13aa9f21
    /css21_dev/html4/reference/filler-text-below-green.htm 27363e8efa7a8550e88a795d9c6e9c9a9f6327cb
    Testing aa3afd43bed732a0c6cdc22b3a71303a13aa9f21 == 27363e8efa7a8550e88a795d9c6e9c9a9f6327cb

emilio committed:
script: Fix a few load related bugs.

This is what was making me hit the new test failures. So turns out that when the
DOMContentLoaded event is fired we fired no messages to the constellation, but
we fired the DOMLoad message from the DocumentProgressHandler, effectively after
having dispatched the Load message from script thread.

This also fixes the possibility of a subframe navigation not blocking the load
event of the parent document, for example.

emilio commented:
@bors-servo: try

bors-servo commented:
:hourglass: Trying commit e4451d0 with merge 590f533...

bors-servo commented:
:broken_heart: Test failed - [linux-rel](http://build.servo.org/builders/linux-rel/builds/2306)

emilio committed:
Some debugging improvements and code style nits across gfx and constellation.

 * Propagate the RUST_LOG env var to the child process
 * Add debug information when a recv() fails in paint thread's select!.

emilio committed:
script: Don't hold the stderr lock while doing sync operations with the constellation.

Otherwise if you enable debug logging, you deadlock.

emilio commented:
@bors-servo: try

bors-servo commented:
:hourglass: Trying commit fe56238 with merge 33abd44...

bors-servo commented:
:broken_heart: Test failed - [linux-rel](http://build.servo.org/builders/linux-rel/builds/2310)

emilio commented:
Ok, seems like all the tests are passing except multi-process ones. Need to figure out how I broke it.

emilio commented:
@bors-servo: try

bors-servo commented:
:hourglass: Trying commit 16b2398 with merge ab160e2...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/2029), [arm32](http://build.servo.org/builders/arm32/builds/2142), [arm64](http://build.servo.org/builders/arm64/builds/2144), [linux-dev](http://build.servo.org/builders/linux-dev/builds/2187), [linux-rel](http://build.servo.org/builders/linux-rel/builds/2311), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/2217), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/2203), [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/2275), [windows-dev](http://build.servo.org/builders/windows-dev/builds/89)

emilio commented:
@samlh Sure, so as an initial way to measure, I just grabbed a twitter feed page, removed scripts and iframes, and loaded it with a release build of servo before and after this patch, profiling with `-p 5`, navigating normally trying to interact with elements to trigger incremental restyle. 

Unfortunately I'm worried this is not a exact measure (even the tiniest difference in interaction could trigger way different code paths), but it might be a good approx measure if done for enough time (these are about 120 incremental restyle events, which are not that many, but probably enough).

Anyway, in that (unexact, but probably sort of relevant) test, the layout numbers are as follows:

### Before patch

```
_category_                          _incremental?_ _iframe?_             _url_                  _mean (ms)_   _median (ms)_      _min (ms)_      _max (ms)_       _events_ 
Compositing                             N/A          N/A                  N/A                        0.6851          0.1803          0.0626        196.0905             559
Layout                                  yes          yes     file:///home/emilio/projects/m         76.6177         55.2281         28.1311       5692.9606             385
Layout                                  no           yes     file:///home/emilio/projects/m        990.0932        990.0932        990.0932        990.0932               1
+ Style Recalc                          yes          yes     file:///home/emilio/projects/m         11.2728         10.4425          3.5714        101.0940             117
+ Style Recalc                          no           yes     file:///home/emilio/projects/m        840.2206        840.2206        840.2206        840.2206               1
| + Text Shaping                        yes          yes     file:///home/emilio/projects/m          0.0583          0.0046          0.0000          2.2492              93
| + Text Shaping                        no           yes     file:///home/emilio/projects/m          2.7505          2.7505          2.7505          2.7505               1
```

### After patch

```
_category_                          _incremental?_ _iframe?_             _url_                  _mean (ms)_   _median (ms)_      _min (ms)_      _max (ms)_       _events_ 
Compositing                             N/A          N/A                  N/A                        0.4669          0.1659          0.0518        131.3618             799
Layout                                  yes          yes     file:///home/emilio/projects/m         74.4345         55.6957         48.5256       6706.9546             399
Layout                                  no           yes     file:///home/emilio/projects/m       1096.8402       1096.8402       1096.8402       1096.8402               1
+ Style Recalc                          yes          yes     file:///home/emilio/projects/m          5.3830          3.6911          0.9647        182.8923             138
+ Style Recalc                          no           yes     file:///home/emilio/projects/m        840.3075        840.3075        840.3075        840.3075               1
| + Text Shaping                        yes          yes     file:///home/emilio/projects/m          0.0483          0.0053          0.0000          2.7329             115
| + Text Shaping                        no           yes     file:///home/emilio/projects/m          2.4944          2.4944          2.4944          2.4944               1
```

Which is more than a 50% improvement on mean incremental style recalc (5.3s vs 11.2s). Again, while this allows some sort of measurement, take these with a *huge* grain of salt, since these are not exact tests. Also, note that this matters even more for the case I was optimizing (stylo), since there we avoid way more FFI calls.

Also, note that Twitter's DOM is not really huge or deep, which was the worst case for the previous system, and that tests with different count of DOM nodes and different depth, or dirtying DOM elements at the top of the document, could make the improvement way more apparent, while this tests only the hovering over elements almost at the bottom of the DOM.

To summarize, the wins here are two:

 1) Not having to traverse all the node for every tiny incremental restyle, including running flow reconstruction, and
 2) Not having to traverse the whole subtree of an element when marking it as dirty.

I've tried to be fair with the data using a somewhat normal page, but some edge cases could make the servo version without this patch *really* slow.

emilio commented:
Ok, so I squashed appropriately and I think I got everything right. I need @jdm or @Ms2ger to review e6958d3947adedda7f3f52f26fae321f0b4d0ad4 and 8c48119f7409ef9a122efcc42872cf0577dccab4, since these are load-related changes. I'm leaving some comments on it, but basically is fixing the messages that go to the constellation after triggering DOMContentReady and window load, to do what they are intended to do.

Basically, with e6958d3947adedda7f3f52f26fae321f0b4d0ad4:

 * The `DOMLoad` message to the constellation goes after triggering `DOMContentLoaded`, instead of after triggering the `load` event on the window.
 * The `LoadComplete` message is sent after actually triggering the window's `load` event (and the appropriate reflows), instead of after enqueuing the runnable to do so. This is what was intended according to the documentation in the definition of the message.
 * The `subframe_loaded` is handled after `LoadComplete` instead of `DOMLoad`, which is what was happening before incidentally, since we sent the `DOMLoad` event after `window.onload`.

Also, the other commit, 8c48119f7409ef9a122efcc42872cf0577dccab4, prevents the parent document to get its blocking loads list empty when an iframe navigates to another page. This was the intended thing according to the comment, but in practice I *think* it didn't happen, and now that layout wasn't necessarily busy when the `LoadComplete` message arrived, started to bit in the reftests "is-ready-to-take-an-image detector". I'm trying to verify if that's what happened actually, or if it was one of the above changes the thing that was hitting me. Also, I'm trying to write a test for that, since it *could* have web-visible behavior (I'm thinking of a parent `window.onload` triggered before the child iframe's).

Also, I filled #12591 about the random multiprocess breakage (that for obvious reasons is no longer in this PR).

emilio commented:
Actually, after looking at it a bit more carefully, I have indeed changed the intended semantics of those messages. I think the actual ones are clearer though. Happy to wait until review.

emilio commented:
To make a bit clearer what my changes attempted, the constellation used the `LoadComplete` message to identify the document as ready (it was used to take a screenshot later).

The problem is the message is sent before the reflow associated with a document load happens. Before, layout was still busy, so the reftests screenshot wasn't taken until it finished, but now it was being taken.

That makes me wonder:

@jdm: what's the reason for the differences between `DOMLoad` and `LoadComplete` messages to the constellation? 

Sounds to me like before a `LoadComplete` arriving implied a `DOMLoad` coming soon. And what we did in `LoadComplete` was somewhat broken for the reasons stated above.

With this patch `DOMLoad` means `DOMContentReady`, but right now it's only used for webdriver stuff that in fact is intended to run in `LoadComplete`, so after fixing it, I don't think we should need two separate messages.

bholley commented on components/style/parallel.rs:
```diff
@@ -63,25 +63,34 @@ fn top_down_dom<N, C>(unsafe_nodes: UnsafeNodeList,
 
+        if !context.should_process(node) {
+            continue;
+        }
```
Check the root in traverse_dom and then remove this check?

bholley commented on components/style/traversal.rs:
```diff
@@ -147,6 +147,12 @@ pub trait DomTraversalContext<N: TNode>  {
+    /// Returns if the node should be processed by the preorder traversal.
+    fn should_process(&self, _node: N) -> bool { true }
+
+    /// Do an action over the child before pushing him to the work queue.
```
"pushing it"

bholley commented on components/style/traversal.rs:
```diff
@@ -245,21 +251,15 @@ pub fn recalc_style_at<'a, N, C>(context: &'a C,
+                unsafe {
+                    node.set_dirty_on_viewport_size_changed();
+                }
+                node.set_descendants_dirty_on_viewport_size_changed();
```
This shouldn't be needed?

bholley commented on ports/geckolib/traversal.rs:
```diff
@@ -35,5 +36,20 @@ impl<'lc, 'ln> DomTraversalContext<GeckoNode<'ln>> for RecalcStyleOnly<'lc> {
     }
 
+    fn should_process(&self, node: GeckoNode<'ln>) -> bool {
+        node.is_dirty() || node.has_dirty_descendants()
```
File a followup to coalesce this to one FFI call?

bholley commented:
First three commits lgtm - have someone look at the dom pieces, and have @pcwalton do a high-level signoff on the parallel traversal piece. Nice find!

pcwalton commented on components/style/parallel.rs:
```diff
@@ -63,25 +63,34 @@ fn top_down_dom<N, C>(unsafe_nodes: UnsafeNodeList,
+        // Possibly enqueue the children.
+        let mut children_to_process = 0isize;
+        for kid in node.children() {
+            context.pre_process_child_hook(node, kid);
```
Could you add a comment here explaining what `pre_process_child_hook` typically does? It'll be important for anyone reading this code to understand the flow of dirty bits.

pcwalton commented:
The approach seems good to me. 👍

emilio committed:
style: Remove a few more unuseful traversals now we can.

emilio commented:
@bors-servo: try

bors-servo commented:
:hourglass: Trying commit 52561ff with merge 2c336b5...

emilio committed:
style: Add a comment about pre_process_children_hook.

emilio commented:
@bors-servo: try

bors-servo commented:
:hourglass: Trying commit a9320fc with merge b858e76...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/2047), [arm32](http://build.servo.org/builders/arm32/builds/2159), [arm64](http://build.servo.org/builders/arm64/builds/2163), [linux-dev](http://build.servo.org/builders/linux-dev/builds/2206), [linux-rel](http://build.servo.org/builders/linux-rel/builds/2330), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/2236), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/2221), [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/2295), [windows-dev](http://build.servo.org/builders/windows-dev/builds/108)

emilio committed:
script: Unify LoadComplete and DOMLoad messages.

See the PR in which this commit landed and also
https://github.com/servo/servo/pull/6415#issuecomment-122294169

emilio committed:
layout: Remove now unused dirty_all_nodes function in layout thread.

jdm commented:
Everything besides the iframe changes look good to me; those are unnecessary because we check for any outstanding loads before queuing the final document load event.

emilio commented:
@bors-servo: r=bholley,jdm,pcwalton

Removed the iframe commit, thanks Josh!

bors-servo commented:
:pushpin: Commit f6043a2 has been approved by `bholley,jdm,pcwalton`

<!-- @bors-servo r=bholley,jdm,pcwalton f6043a23942816989c51798c069c11e8e8726882 -->

bholley commented:
@bors-servo p=1

bors-servo commented:
:hourglass: Testing commit f6043a2 with merge 8d1ddba...

bors-servo commented:
:broken_heart: Test failed - [linux-rel](http://build.servo.org/builders/linux-rel/builds/2359)

bholley commented:
@bors-servo retry

bholley commented:
Tests with unexpected results:
  ▶ Unexpected subtest result in /html/browsers/history/the-history-interface/001.html:
  │ FAIL [expected PASS] pushState should not actually load the new URL
  │   → assert_true: expected true got undefined
  │ 
  │ tests12/<@http://web-platform.test:8000/html/browsers/history/the-history-interface/001.html:283:13
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1402:20
  │ test@http://web-platform.test:8000/resources/testharness.js:500:9
  └ tests12@http://web-platform.test:8000/html/browsers/history/the-history-interface/001.html:282:9

#12580 #12625

bors-servo commented:
:zap: Previous build results for [android](http://build.servo.org/builders/android/builds/2070), [arm32](http://build.servo.org/builders/arm32/builds/2185), [arm64](http://build.servo.org/builders/arm64/builds/2187), [linux-dev](http://build.servo.org/builders/linux-dev/builds/2234), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/2266), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/2250), [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/2323), [windows-dev](http://build.servo.org/builders/windows-dev/builds/133) are reusable. Rebuilding only [linux-rel](http://build.servo.org/builders/linux-rel/builds/2359)...

bors-servo commented:
:broken_heart: Test failed - [linux-rel](http://build.servo.org/builders/linux-rel/builds/2361)

emilio commented:
@bors-servo: retry

 * #12625 (the other test).

bors-servo commented:
:hourglass: Testing commit f6043a2 with merge 944d371...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/2073), [arm32](http://build.servo.org/builders/arm32/builds/2188), [arm64](http://build.servo.org/builders/arm64/builds/2190), [linux-dev](http://build.servo.org/builders/linux-dev/builds/2237), [linux-rel](http://build.servo.org/builders/linux-rel/builds/2363), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/2269), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/2253), [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/2326), [windows-dev](http://build.servo.org/builders/windows-dev/builds/136)
